### ISSUE TITLE
Support search result classes different from Specification query root [DATAJPA-51]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-jpa#487]
**[Stevo Slavić](https://jira.spring.io/secure/ViewProfile.jspa?name=sslavic)** opened **[DATAJPA-51](https://jira.spring.io/browse/DATAJPA-51?redirect=false)** and commented

Currently Specification API does not support search to return anything other than instance of query root (or some collection of root) JPA entity as result. But JPA itself supports result to be different class from query root and it typically isn't an entity bean class (see [here](http://www.objectdb.com/java/jpa/query/jpql/select#Result_Classes_Constructor_Expressions_) and [here](http://www.objectdb.com/java/jpa/query/jpql/select#CriteriaBuilder-s_construct_)).

Please add support for complex queries with search result classes different from Specification query root

---

**Affects:** 1.0 M2

42 votes, 25 watchers

**Comment by yasslkb** [comment]
we need this

**Comment by ilkou** [comment]
in Hexagonal Architecture for example we have a `domain` class and a `jpaEntity` class;

- the domain class could be:

```java
class User {
   private String id;
   private String fullname;
   private String phone;
   private String stuff;
   ...
}
```
and 
```java
class Wallet {
   private String id;
   private String balance;
   private User user; // we could have a constructor (lite modal) with only fullname / id
   ...
}
```

- the corresponding jpa entities could be:

```java
class UserJpaEntity {
   private String id;
   private String fullname;
   private String phone;
   private String stuff;
   ...
}
```
and 
```java
class WalletJpaEntity {
   private String id;
   private String balance;
   private String userId;
   ...
}
```

now due to Hexagonal Architecture, we can't just use @ OneToOne or such annotations to link jpa entities and call it a day, I think..

Let's tackle the problem;

To fetch List/Page of Wallets for example, with their Users, one might fetch all wallets then loop through them and fetch the User, and that sounds like [n+1](https://spring.io/blog/2023/08/31/this-is-the-beginning-of-the-end-of-the-n-1-problem-introducing-single-query) problem 🤔

or we can simply make a custom Wallet constructor and use it to map the result with @ Query, like the following:
```java
@Query("""
        SELECT new io.spring.demo.Wallet(w.id, w.balance, u.id, u.fullname) FROM WalletJpaEntity w
        INNER JOIN UserJpaEntity u ON u.id = w.userId
        """)
Page<Wallet> findWallets(Pageable pageable);
```
and this works 🎉 

But...

For filtering for example, we would need to use Specification API (that's why I'm commenting on this Issue), like this:
```java
@Query("""
        SELECT new io.spring.demo.Wallet(w.id, w.balanace, u.id, u.fullname) FROM WalletJpaEntity w
        INNER JOIN UserJpaEntity u ON u.id = w.userId
        """)
Page<Wallet> findWallets(Specification<WalletJpaEntity> spec, Pageable pageable);
```
which doesn't work currently....

any work around this to assure we can have dynamic filters + efficient queries with joins (custom result mapping) at the same time ?

PS: A [StackOverFlow](https://stackoverflow.com/questions/62567283/spring-jpa-specification-api-with-custom-query-and-custom-response-object-is-th) question with same issue

**Comment by mp911de** [comment]
This will be fixed via #2327
